File: api/home-box-landing/HomeBoxLanding.Api/Features/Columns/ColumnMapper.cs
using HomeBoxLanding.Api.Features.Columns.Types;
using HomeBoxLanding.Api.Features.Folders;
using HomeBoxLanding.Api.Features.Links;

namespace HomeBoxLanding.Api.Features.Columns;

public class ColumnMapper
{
    public static Column Map(ColumnRecord record)
    {
        return new Column
        {
            Identifier = record.Identifier,
            Name = record.Name,
            SortOrder = record.SortOrder,
            Icon = record.Icon,
            // Links held by a folder are returned on that folder, never loose on the column.
            Links = record.Links
                .Where(x => x.FolderIdentifier == null)
                .OrderBy(x => x.SortOrder)
                .Select(LinkMapper.Map)
                .ToList(),
            Folders = record.Folders
                .OrderBy(x => x.SortOrder)
                .Select(FolderMapper.Map)
                .ToList()
        };
    }
}

File: src/services/link-service/column.mapper.ts
import { LinkMapper } from './link.mapper';
import { FolderMapper } from './folder.mapper';
import { IColumn } from './types/column.type';

export class ColumnMapper {

    public static map(response: any): Array<IColumn> {
        return response.map((column: any) => ColumnMapper.mapSingle(column));
    }

    public static mapSingle(column: any): IColumn {
        return {
            identifier: column.Identifier,
            name: column.Name,
            sortOrder: column.SortOrder,
            icon: column.Icon,
            links: column.Links.map((link: any) => LinkMapper.mapSingle(link)),
            folders: FolderMapper.map(column.Folders)
        };
    }

    public static mapToApi(columns: Array<IColumn>): any {
        return columns.map((column: any) => ColumnMapper.mapToApiSingle(column));
    }

    public static mapToApiSingle(column: IColumn): any {
        return {
            Identifier: column.identifier,
            Name: column.name,
            SortOrder: column.sortOrder,
            Icon: column.icon,
            Links: LinkMapper.mapToApi(column.links),
            Folders: FolderMapper.mapToApi(column.folders)
        };
    }
}